src/Microsoft.Identity.Client/Azure/ChainedTokenProvider.cs

```diff
@@ -0,0 +1,55 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Microsoft.Identity.Client.Azure
+{
+    /// <summary>
+    /// ChainedTokenProvider creates an ordered list of probes that will determine if the environment will allow it to build
+    /// a ICredentialProvider. The ICredentialProvider is then used to generate authentication credentials for the
+    /// consumer.
+    /// </summary>
+    public class ChainedTokenProvider : ITokenProvider
+    {
+        private readonly IEnumerable<IProbe> _probes;
+
+        /// <summary>
+        /// Create an instance of a ChainedTokenProvider providing a list of IProbes which will be executed in order to create a ICredentialProvider
+        /// </summary>
+        /// <param name="probes">probes to be executed in order to create a ICredentialProvider</param>
+        /// <exception cref="ArgumentException">throws if no probes were provided or if no probe is able to build a ICredentialProvider</exception>
+        public ChainedTokenProvider(IEnumerable<IProbe> probes)
+        {
+            if (probes == null || probes.Count() < 1)
+            {
+                throw new ArgumentException("must provide 1 or more IProbes");
+            }
+            _probes = probes;
+        }
+
+        /// <summary>
+        ///     GetTokenAsync returns a token for a given set of scopes
+        /// </summary>
+        /// <param name="scopes">Scopes requested to access a protected API</param>
+        /// <returns>A token with expiration</returns>
+        public async Task<IToken> GetTokenAsync(IEnumerable<string> scopes = null)
+        {
+            ITokenProvider provider = null;
+            foreach (var probe in _probes)
+            {
+                if (await probe.AvailableAsync().ConfigureAwait(false))
+                {
+                    provider = await probe.ProviderAsync().ConfigureAwait(false);
```

Comment:
Do we not exist loop once provider has been found?

Reply:
Definitely should. Good catch.

Reply:
updated